### PR TITLE
Implement reverse geolocation via /findNearbyPlaceNameJSON endpoint

### PR DESCRIPTION
**Pull request by mfilej**
Adds `reverse_find` and `reverse_search` endpoints to client.

The refactoring should allow us to handle any errors returned by API in a sane manner.

### CODE REVIEW

File: lib/gemonames.rb [hunk 1]
```diff
@@ -36,17 +37,43 @@ def initialize(connection)
...
+
+    def reverse_search(latitude:, longitude:, limit: 10)
+      perform_search :find_nearby_place_name,
+        lat: latitude, lng: longitude, maxRows: limit
+    end
+
+    def reverse_find(latitude:, longitude:)
+      perform_find :find_nearby_place_name,
+        lat: latitude, lng: longitude
+    end
```
Comment (mfilej) [inline lib/gemonames.rb]: I guess I should comment on the above changes:

I tried to achieve a shape where it is easily visible which endpoints are called on the geonames API and which parameters are passed exactly (that's why I changed to the ugly `maxRows` for example).

File: lib/gemonames.rb [hunk 2]
```diff
@@ -36,17 +37,43 @@ def initialize(connection)
...
+
+    private
+
+    def perform_search(endpoint, **args)
```
Comment (tomekw) [inline lib/gemonames.rb]: Keyword arg for `endpoint`? `perform_search(:search, {})` reads weird :)

File: lib/gemonames.rb [hunk 3]
```diff
@@ -36,17 +37,43 @@ def initialize(connection)
     end
 
     def search(query, country_code:, limit: 10)
-      perform_search_request(
-        query: query, country_code: country_code, max_rows: limit
-      ).fetch("geonames").map { |result|
+      perform_search :search,
```
Comment (tomekw) [inline lib/gemonames.rb]: Parentheses around arguments (everywhere) :)

File: lib/gemonames.rb [hunk 4]
```diff
@@ -36,17 +37,43 @@ def initialize(connection)
...
         wrap_in_search_result(result)
       }
     end
 
-    def find(query, country_code:)
-      results = perform_search_request(
-        query: query, country_code: country_code, max_rows: 1
-      ).fetch("geonames")
+    def perform_find(endpoint, **args)
+      args = args.merge(maxRows: 1)
```
Comment (tomekw) [inline lib/gemonames.rb]: Maybe just inline it on line 75?

File: lib/gemonames.rb [hunk 5]
```diff
@@ -36,17 +37,43 @@ def initialize(connection)
...
+    end
+
+    private
+
+    def perform_search(endpoint, **args)
+      results = extract_payload(
+        WebServices.public_send(endpoint, connection, **args)
+      )
+
+      results.map { |result|
```
Comment (tomekw) [inline lib/gemonames.rb]: One liner?

File: lib/gemonames.rb [hunk 6]
```diff
@@ -36,17 +37,43 @@ def initialize(connection)
...
+
+    def reverse_find(latitude:, longitude:)
+      perform_find :find_nearby_place_name,
+        lat: latitude, lng: longitude
+    end
+
+    private
+
+    def perform_search(endpoint, **args)
+      results = extract_payload(
```
Comment (tomekw) [inline lib/gemonames.rb]: `results` variable not needed? Just chain it with `map`?

File: lib/gemonames.rb (hunk 3 above)
Comment (mfilej) [inline lib/gemonames.rb]: I'd prefer to skip them as they would add 2 additional lines to each method.

File: lib/gemonames.rb (hunk 2 above)
Comment (mfilej) [inline lib/gemonames.rb]: True. :smile: Maybe I can think of another name.

**Comment by mfilej** [comment]
@tomekw re: inlining - I left the code in this shape so that it will be easier to refactor.

final form

**Comment by mfilej** [comment]
@tomekw I think the refactoring addressed all your comments (except parentheses).

**Comment by RadekMolenda** [comment]
:star:

**Comment by tomekw** [comment]
:star:

**Pull request merged**